### ISSUE TITLE
threadpool segfaults on interrupt  CTRL_C

### ISSUE THREAD

**Issue by alexeadem** [warmcat/libwebsockets#2226]
Hello,

I'm trying to build a ws proxy that requires multiple proxy clients request to be run as one
```
single client ws => ws proxy => ws client request/response #1
                                                 ws client request/response #2
                                                 ...
```
I got this working with `pss` per session `struct` but the issue is when the client disconnect, the remaining cmds will fail as the `pss` struct will be destroyed on close. I understand this is expected and I was looking for other options.

The other option is to use vhd (per host) but every new client connection behind the proxy data  vhd will be overwritten. 

Then I found the threadpool minimal code which should continue a task even if the browser/client disconnects.

I ran a couple of tests with 

`minimal-examples/client-server/minimal-ws-proxy/minimal-ws-proxy.c`

but it segfaults in Valgrind after the service is interrupted. 

```
==1820287== Thread 2:
==1820287== Invalid read of size 8
==1820287==    at 0x41E7F1: lws_get_context (wsi.c:875)
==1820287==    by 0x40DBB0: lws_threadpool_worker (threadpool.c:662)
==1820287==    by 0x487A431: start_thread (in /usr/lib64/libpthread-2.31.so)
==1820287==    by 0x4E9A6D2: clone (in /usr/lib64/libc-2.31.so)
==1820287==  Address 0x4f972f0 is 0 bytes inside a block of size 1,160 free'd
==1820287==    at 0x483B9F5: free (vg_replace_malloc.c:538)
==1820287==    by 0x44DBB5: _realloc (alloc.c:144)
==1820287==    by 0x44DBEC: lws_realloc (alloc.c:154)
==1820287==    by 0x414163: __lws_free_wsi (close.c:194)
==1820287==    by 0x4155CD: __lws_close_free_wsi_final (close.c:826)
==1820287==    by 0x41541A: __lws_close_free_wsi (close.c:778)
==1820287==    by 0x415627: lws_close_free_wsi (close.c:836)
==1820287==    by 0x408516: lws_context_destroy (context.c:1723)
==1820287==    by 0x4067D3: main (minimal-ws-server-threadpool.c:132)
==1820287==  Block was alloc'd at
==1820287==    at 0x483A747: malloc (vg_replace_malloc.c:306)
==1820287==    by 0x483CD32: realloc (vg_replace_malloc.c:834)
==1820287==    by 0x44DB5B: _realloc (alloc.c:133)
==1820287==    by 0x44DC17: lws_zalloc (alloc.c:159)
==1820287==    by 0x41D5BB: __lws_wsi_create_with_role (wsi.c:221)
==1820287==    by 0x450FD4: lws_create_new_server_wsi (adopt.c:62)
==1820287==    by 0x45121F: lws_adopt_descriptor_vhost1 (adopt.c:136)
==1820287==    by 0x4518FC: lws_adopt_descriptor_vhost_via_info (adopt.c:510)
==1820287==    by 0x451796: lws_adopt_descriptor_vhost (adopt.c:480)
==1820287==    by 0x440B23: rops_handle_POLLIN_listen (ops-listen.c:147)
==1820287==    by 0x41C609: lws_service_fd_tsi (service.c:735)
==1820287==    by 0x445632: _lws_plat_service_forced_tsi (unix-service.c:49)

```

It looks like by interrupting the service with CTRL_C we don't wait for the threads to terminate effectively destroying the context they use.

```
	while (!interrupted)
		if (lws_service(context, 0))
			interrupted = 1;

	lws_context_destroy(context);

```

Version:

v4.1.0-292-g0455c66e

Not sure if that's the right way to go for my use case but I was looking for a ws proxy that is independent of the main process and can continue running multiple proxy calls even if the user has disconnected.

**Comment by lws-team** [comment]
Yes the threadpool example doesn't wait for completion of ongoing threads if you terminate it.  Ongoing threads may live forever, user code can trap SIGINT, set a flag to reject new connections, ask threads to terminate next time they check in and finally exit the loop when the last one goes down.  The "minimal" example takes the approach to just exit unconditionally since the process is going down.

Threadpool itself is robust and valgrind-clean, it's used in gitohashi which is extremely stable.

Anyway you do not want threadpool from your description, it's needed for managing thread-blocking work associated with a wsi, if I understand it you, you instead have a sequence of async tasks associated with a wsi, that can all be done on the event loop.

The right pattern for this is allocate a new 'connection' object (like a struct my_proxying_connection) when the incoming wsi connection is ESTABLISHED, and mark the incoming wsi with a pointer to it using `lws_set_opaque_user_data()`.  In your connection object, it keeps track of related wsi pointers, and things like overall proxying state.  Any new outgoing connections created on behalf of this proxy connection also have their wsi opaque pointers set to the related connection object and point to their wsi in there, so any related wsi can get to the connection object and get wsi pointers for the other participants.  When those outbound wsi close, they should remove or set to NULL their reference in the connection object.  If they see they were the last wsi referenced by the connection object, they can destroy it.

The inbound ws connection _CLOSE callback should also remove or set NULL any reference to its wsi in the connection object, and because his going down implies any other connections started on his behalf are now cancelled, either start the close flow on any still existing subordinate wsi still listed in the connection object, or if none, destroy the connection object.  This handles any situation of wsi closing order cleanly.

The proxy connection object can manage as much sequencing as you like for whatever the lifetime of the proxy action should be, all async on the event loop, no need for locking or threads (and no difficulty at shutdown since it's just wsi closing).

**Comment by lws-team** [comment]
Here's an example that demonstrates what I wrote above, for the case where the proxy is a ws server making a raw socket onward client connection.

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/ws-server/minimal-ws-raw-proxy?h=main

I am not 100% sure what you were trying to do, but eg, this example does not have any pss or vhd, just the common proxy connection object while either side of the proxy is still alive.

**Comment by alexeadem** [comment]
Thanks for the guidance. The example you provided solves part of the challenge. In the example you provided 

```
                if (pc->pending_msg_to_ws.count) {
                        /*
                         * Yes, let him get on with trying to send
                         * the remaining pieces... but put a time limit
                         * on how hard he will try now the raw part is
                         * disappearing... give him 3s
                         */
                        lws_set_timeout(pc->wsi_ws,
                                PENDING_TIMEOUT_KILLED_BY_PROXY_CLIENT_CLOSE, 3);
                        break;
                }

``
you wait for the current proxy to finish.

In my case, I need to call to  writable again
``
		lws_callback_on_writable(pss->wsi);
or
		//lws_callback_on_writable_all_protocol(vhd->context, vhd->protocol);

``
for the next proxy API call to run. 

The challenge is that the wsi is gone and the proxy must go on in the writable loop to finalize the API calls

Basically, the proxy needs to keep going once started and even if everyone disconnected.

Example: 
the client connects and starts proxy API call 1, 2, 4

every API call will start with:

case LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP:

and it will end with 

case LWS_CALLBACK_CLOSED_CLIENT_HTTP

then fo to here to write output and run the next command

case LWS_CALLBACK_SERVER_WRITEABLE:

and everything starts again

case LWS_CALLBACK_ESTABLISHED_CLIENT_HTTP:
...

until all calls are done


A beautiful loop that works perfectly as long as there a wsi to hold the data but it stops as soon and it disappears

**Comment by lws-team** [comment]
The example is just one example of arbitrary proxy behaviours that are valid, it's designed to show how to implement the proxy conn object scheme cleanly.  If it's not exactly your scenario, well, you can adapt it, it's not trying to do your job for you.

> Basically, the proxy needs to keep going once started and even if everyone disconnected.

I don't think you mean "everyone", because then it's just leaking if they are immortal.  You have to take care of broken attempts and put a time limit on what it's trying to do, even if it becomes detached from the original ws wsi.

The outgoing client connections themselves do not depend on the ws wsi, they can continue or spawn new connections or whatever.  The point of the proxy example is that the original, triggering ws connection contributes creating the proxy conn object they all participate in, which can perfectly well outlive the original ws connection.

If you focus on the proxy conn object instead of the ws, there is nothing stopping you doing multi-client sequencing on that even if the original ws is long dead, if you adapt the example accordingly.

**Comment by alexeadem** [comment]
will do. Thank you for all your help. Alex
